Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs

```diff
@@ -291,117 +293,154 @@ private Point GetRealCaretVisualPoint()
-            return new FormattedText("^", CultureInfo.CurrentUICulture, FlowDirection.RightToLeft, textRunProperties.Typeface, textRunProperties.FontRenderingEmSize, Brushes.Black);
+            return new FormattedText("A", CultureInfo.CurrentUICulture, FlowDirection.LeftToRight, textRunProperties.Typeface, textRunProperties.FontRenderingEmSize, Brushes.Black);
         }
 
         /// <summary>
         /// Calculate the dimensions of the caret
         /// </summary>
-        private Size CalculateCaretSize()
+        private Size CalculateCaretSize(out string caretCharacter)
         {
+            caretCharacter = "";
             var defaultWidth = FormattedText.Width;
 
             var caret = _textView.Caret;
             var line = caret.ContainingTextViewLine;
             var width = defaultWidth;
-            var height = caret.Height;
+            var height = line.TextHeight;
+
             if (IsRealCaretVisible)
             {
-                // Get the size of the character to which we need to paint the caret.  Special case
-                // tab here because it's too big.  When there is a tab we use the default height
-                // and width
+                // Get the size of the character to which we need to paint
+                // the caret.  Special case tab here because it's too big.
+                // When there is a tab we use the default height and width.
                 var point = caret.Position.BufferPosition;
-                if (point.Position >= _textView.TextSnapshot.Length || point.GetChar() == '\t')
-                {
-                    width = defaultWidth;
-                }
-                else
+                if (point.Position < _textView.TextSnapshot.Length)
                 {
-                    var bounds = line.GetCharacterBounds(point);
-                    if (_controlCharUtil.IsDisplayControlChar(point.GetChar()))
-                    {
-                        width = defaultWidth;
-                    }
-                    else
+                    var pointCharacter = point.GetChar();
+                    if (pointCharacter != '\t'
+                        && !_controlCharUtil.IsDisplayControlChar(pointCharacter))
                     {
-                        width = bounds.Width;
-                        height = bounds.TextHeight;
+                        // Handle surrogate pairs.
+                        if (Char.IsHighSurrogate(pointCharacter)
+                            && point.Position < _textView.TextSnapshot.Length - 1
+                            && Char.IsLowSurrogate(point.Add(1).GetChar()))
+                        {
+                            caretCharacter = new SnapshotSpan(point, 2).GetText();
+                        }
+                        else
+                        {
+                            caretCharacter = pointCharacter.ToString();
+                        }
+                        width = line.GetCharacterBounds(point).Width;

```

Comment:
Why does the width of a single point get used here when there is a surrogate pair?

Reply:
The `GetCharacterBounds` API effectively means "get the bounds of the character to the right of the caret, taking into account any surrogate pairs". In other words, the text view line itself it well aware of the surrogate pair.